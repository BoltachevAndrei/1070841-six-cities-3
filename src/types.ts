export enum SortTypeList {
  POPULAR = `Popular`,
  LOW_TO_HIGH = `Price: low to high`,
  HIGH_TO_LOW = `Price: high to low`,
  TOP_RATED_FIRST = `Top rated first`
};

export enum AuthorizationStatus {
  AUTH = `AUTH`,
  NO_AUTH = `NO_AUTH`
};

export type City = {
  location: {
    latitude: number,
    longitude: number,
    zoom: number
  },
  name: string
};

export type User = {
  id: number,
  name: string,
  avatar: string,
  isSuper: boolean,
  email?: string
};

export type Offer = {
  id: number,
  city: City,
  isPremium: boolean,
  images: Array<string>,
  price: number,
  isBookmarked: boolean,
  rating: number,
  title: string,
  description: string,
  features: {
    entire: string,
    bedrooms: string,
    adults: string
  },
  goodsInside: Array<string>,
  host: User,
  coordinates: Array<number>,
  zoom: number,
  previewImage: string
};

export type Comment = {
  text: string,
  date: Date,
  id: number,
  rating: number,
  user: User
};
