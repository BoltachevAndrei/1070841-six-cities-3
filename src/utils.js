import {SortTypeList} from './types';

export const capitalizeString = (str) => str[0].toUpperCase() + str.slice(1);

export const extend = (object1, object2) => Object.assign({}, object1, object2);

export const getCities = (data) => Array.from(new Set(data.map((element) => element.city)).values());

export const getOffersByCity = (offers, city) => offers.filter((offer) => offer.city === city);

export const sortOffers = (offers, sortType) => {
  switch (sortType) {
    case SortTypeList.HIGH_TO_LOW:
      return offers.sort((prev, next) => next.price - prev.price);
    case SortTypeList.LOW_TO_HIGH:
      return offers.sort((prev, next) => prev.price - next.price);
    case SortTypeList.TOP_RATED_FIRST:
      return offers.sort((prev, next) => next.rating - prev.rating);
  }
  return offers;
};

export const getCitiesList = (offers) => {
  const citiesNames = Array.from(new Set(offers.map((element) => element.city.name)).values());
  const citiesData = Array.from(offers.values()).map((offer) => offer.city);
  const result = citiesNames.map((name) => citiesData.filter((city) => city.name === name)[0]);
  return result;
};

export const doNothing = () => {
  // do nothing
};
