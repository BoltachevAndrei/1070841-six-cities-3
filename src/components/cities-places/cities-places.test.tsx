import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Router} from 'react-router-dom';
import history from '../../history';
import CitiesPlaces from './cities-places';
import {SortTypeList} from '../../types';
import {doNothing} from '../../utils';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);

jest.mock(`../map/map.tsx`);

const offers = [
  {
    id: 111,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    isPremium: true,
    images: [
      `img/amsterdam.jpg`
    ],
    price: 222,
    isBookmarked: true,
    rating: 1,
    title: `Test title 1`,
    description: `Test description`,
    features: {
      entire: `Test type 1`,
      bedrooms: `33`,
      adults: `55`
    },
    goodsInside: [
      `MiniBar`
    ],
    host: {
      id: 1,
      name: `Ivan`,
      avatar: `img/avatar.svg`,
      isSuper: true
    },
    previewImage: `img/amsterdam.jpg`,
    coordinates: [
      52.370216,
      4.895168
    ],
    zoom: 12
  },
  {
    id: 333,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    isPremium: false,
    images: [
      `img/amsterdam@2x.jpg`
    ],
    price: 444,
    isBookmarked: false,
    rating: 5,
    title: `Test title 2`,
    description: `Test description`,
    features: {
      entire: `Test type 2`,
      bedrooms: `33`,
      adults: `55`
    },
    goodsInside: [
      `MiniBar`
    ],
    host: {
      id: 1,
      name: `Ivan`,
      avatar: `img/avatar.svg`,
      isSuper: true
    },
    previewImage: `img/amsterdam@2x.jpg`,
    coordinates: [
      52.370216,
      4.895168
    ],
    zoom: 12
  }
];

const activeCity = {
  name: `Amsterdam`,
  location: {
    latitude: 52.370216,
    longitude: 4.895168,
    zoom: 10
  }
};

const placesCount = 111;

const card = 0;

const sortType = SortTypeList.POPULAR;

it(`CitiesPlaces renders correctly`, () => {
  const store = mockStore({
    APP_STATE: {
      sortType,
      isSortListOpened: false
    }
  });
  const tree = renderer
    .create(
        <Router
          history={history}
        >
          <Provider store={store}>
            <CitiesPlaces
              offers={offers}
              activeCity={activeCity}
              placesCount={placesCount}
              card={card}
              sortType={sortType}
              onPlaceCardClick={doNothing}
              onPlaceCardMouseOver={doNothing}
              onPlaceCardMouseLeave={doNothing}
            />
          </Provider>
        </Router>
    )
  .toJSON();
  expect(tree).toMatchSnapshot();
});
