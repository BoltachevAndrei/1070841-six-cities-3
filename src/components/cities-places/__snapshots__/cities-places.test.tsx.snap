// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesPlaces renders correctly 1`] = `
<div
  className="cities__places-container container"
>
  <section
    className="cities__places places"
  >
    <h2
      className="visually-hidden"
    >
      Places
    </h2>
    <b
      className="places__found"
    >
      111
       places to stay in 
      Amsterdam
    </b>
    <form
      action="#"
      className="places__sorting"
      method="get"
    >
      <span
        className="places__sorting-caption"
      >
        Sort by
      </span>
      <span
        className="places__sorting-type"
        onClick={[Function]}
        tabIndex={0}
      >
        Popular
        <svg
          className="places__sorting-arrow"
          height="4"
          width="7"
        >
          <use
            xlinkHref="#icon-arrow-select"
          />
        </svg>
      </span>
      <ul
        className="places__options places__options--custom "
      >
        <li
          className="places__option places__option--active"
          onClick={[Function]}
          tabIndex={0}
        >
          Popular
        </li>
        <li
          className="places__option"
          onClick={[Function]}
          tabIndex={0}
        >
          Price: low to high
        </li>
        <li
          className="places__option"
          onClick={[Function]}
          tabIndex={0}
        >
          Price: high to low
        </li>
        <li
          className="places__option"
          onClick={[Function]}
          tabIndex={0}
        >
          Top rated first
        </li>
      </ul>
    </form>
    <div
      className="cities__places-list places__list tabs__content"
    >
      <article
        className="cities__place-card place-card"
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="place-card__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="cities__image-wrapper place-card__image-wrapper"
        >
          <a
            href="#"
          >
            <img
              alt="Place image"
              className="place-card__image"
              height={200}
              src="img/amsterdam.jpg"
              width={260}
            />
          </a>
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                222
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="place-card__bookmark-button place-card__bookmark-button--active button"
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="19"
                width="18"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                In bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "20%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                1
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
            onClick={[Function]}
          >
            <a
              href="/offer/111"
              onClick={[Function]}
            >
              Test title 1
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            Test type 1
          </p>
        </div>
      </article>
      <article
        className="cities__place-card place-card"
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        
        <div
          className="cities__image-wrapper place-card__image-wrapper"
        >
          <a
            href="#"
          >
            <img
              alt="Place image"
              className="place-card__image"
              height={200}
              src="img/amsterdam@2x.jpg"
              width={260}
            />
          </a>
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                444
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="place-card__bookmark-button button"
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="19"
                width="18"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                5
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
            onClick={[Function]}
          >
            <a
              href="/offer/333"
              onClick={[Function]}
            >
              Test title 2
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            Test type 2
          </p>
        </div>
      </article>
    </div>
  </section>
  <div
    className="cities__right-section"
  >
    <section
      className="cities__map map"
    />
  </div>
</div>
`;
